### ISSUE TITLE
Viewing/Removing Relationships in Edit Item is difficult for entities with >20 relationships

### ISSUE THREAD

**Issue by tdonohue** [DSpace/dspace-angular#1148]
**Describe the bug**
Discovered while testing #1139 

In the "Edit Item" screen, "Relationships" tab, there are two related issues which make managing >20 relationships difficult.  I've listed these in one ticket, as a single solution _might_ be "good enough" for now.

**[Bug 1]** When adding a new relationship, the (popup) modal window works overall.  However, it doesn't show which entities already have an existing relationship. So, the "Current Selection" tab in the modal always says zero, and every entity in the system is listed as unchecked. This means that you can attempt to add a new relationship to an Entity which _already has an existing relationship_ as there's no way to tell in the popup which ones are already related. (If you attempt to add a duplicate relationship, nothing happens though...the UI knows that one already exists.)

**[Bug 2]** When an Entity has a large number of relationships, it becomes impossible to see them all on the "Relationships" tab. Either the "Relationships" tab needs pagination, or the "Current Selection" tab in the popup should list all existing relationships with pagination. Currently, if there's more than 20 relationships, only the first 20 relationships appear. So, it's impossible to see (or delete) any relationship after the first 20.

**To Reproduce**
1. Login as an Admin
2. Edit an Entity (Item), preferably one with a lot of relationships (>20) like a Person with many Publications. An example on the demo site is Cameron Simmons: https://demo7.dspace.org/entities/person/b1b2c768-bda1-448a-a073-fc541e8b24d9
3. Go to the "Relationships" tab.
4. You'll only see the first 20 related entities listed. There's no way to see the others (Bug 2 above)
5. Click "Add" to see the popup modal for adding a new relationship.  You'll see all existing relationship are not listed (Bug 1 above)... and the "Current Selection" tab says zero.

**Expected behavior**
At the very least, the "Current Selection" tab should work, and this should allow you to also deselect/remove relationships.

Ideally, we'd also have pagination on the "Relationships" tab. But, this could be worked around if the "Current Selection" tab was the recommended approach for removing existing relationships.

**Comment by tdonohue** [comment]
@artlowel : Assigning this to your team to investigate & estimate.  As noted above, I think there might be two possible solutions here...but, I'd appreciate your thoughts/ideas on which direction is better for 7.0 vs later.

**Comment by tdonohue** [comment]
Splitting this ticket into two, as, after further review, I think the two bugs listed in this ticket may need to have separate priorities:

1. I'll leave this ticket as `medium`, but it'll only be about **adding** new relationships (which "works" for now, but it's clunky and needs to be improved to filter out existing relationships).  
2. The second bug around pagination I'm splitting out separately as a `high priority` task, as it means some relationships are unable to be seen (and therefore cannot be removed).  This ideally should be fixed by 7.0 (and I've already talked with @artlowel about doing so).

**Comment by artlowel** [comment]
There is a fundamental problem with the way we check if search results are already related to the item being edited in this lookup. The UI currently needs to:
- get _all_ relationships for the item being edited, 
- then retrieve _all_ the items on both side of the relationship (because we can't know whether the current item is on the left or right side until we fetch it), 
- make a list of the related entities, and pass that to the lookup modal
- and then, when we search for other entities, we check if they are already in that selected list in order to determine if we should show it as selected or not

This is problematic because any time the UI has to get a complete list of a resource in order to do some check, we have to disregard pagination and get the advantages that come with it.

I'd propose we handle this on the rest side, using projections. When the UI searches for new entities to create a possible relationship with, we provide a 2 additional params:
- projection=isRelatedTo
- isRelatedTo=${relationship type}=${item uuid}

E.g. `/api/discover/search/objects?query=test&isRelatedTo=isPublicationOfAuthor=b1b2c768-bda1-448a-a073-fc541e8b24d9&projection=isRelatedTo`

Rest would then check for each ItemRest in the result set whether it has a relationship of the given type to the given item and add the property isRelated: true or false to that Item

Similarly when the UI gets a relationship object we usually have to check if a given item is on the left or right side. The only way to do that is by retrieving both sides and checking. If we add a projection that specifies on which side the given item is, we can increase performance.

The same thing goes for retrieving a relationshiptype, you usually need to know on which side the given entity type is, so again a projection that provides that info would be great.

Note that it is important that this projection info becomes part of the self link of these objects. As they affect the content of  the response it should reflect in the self link.

Finally the edit relationships tab starts by retrieving all possible relationshiptypes, and then filtering out the ones that apply to the current item. So a search endpoint that allows you to retrieve a paginated list of relationshiptypes for a given entity type would allow us to get rid of yet another infinite-pagesize call in the UI

So in summary. I propose we add the following to the rest side
- A projection for items, to indicate whether the item is related to the given item with the given relationship type
- A projection for relationships, to indicate on which side of the relationship the given item resides
- A projection for relationshiptypes, to indicate on which side of the relationshiptype the given entity type resides
- A search endpoint for relationshiptypes, to search by entity type

To rewrite the edit relationships tab to make use of those new projections and endpoints would take an estimated **3 days**

**Comment by tdonohue** [comment]
@artlowel : Overall, this sounds reasonable.  I would recommend though that we start with a RESTContract, as that might more easily describe exactly how these projections would work.  These projections likely should be described under the endpoint where they exist...so, for example, updating the Search Results contract to describe the optional `isRelatedTo` projection which can be used there.

For now, I'll assign this to you and @benbosman & add the estimate.  But, I would recommend we see the proposed Contract sooner rather than later, as it may generate discussion that could improve/enhance the implementation. Thanks!
